### PR TITLE
Some cleanups around LE

### PR DESCRIPTION
**Pull request by vagababov**
- make funcs and types not expected to be used outside private (we can always make them public, should the need arise)
- add some more checks in tests
- fix some strings

/assign @yanweiguo @mattmoor

### CODE REVIEW

File: leaderelection/config.go [hunk 1]
```diff
@@ -127,25 +129,37 @@ type ComponentConfig struct {
...
+	ssName  string
+	ordinal int
 }
 
-func NewStatefulSetConfig() (*StatefulSetConfig, error) {
-	var ssc StatefulSetConfig
-	ssn, _, err := ParseControllerOrdinal()
-	if err != nil {
-		return nil, err
+func (ssID *statefulSetID) Decode(v string) error {
```
Comment (mattmoor) [inline leaderelection/config.go:138]: Wanna add `var _ envconfig.Decoder = (*statefulSetID)(nil)`?

File: leaderelection/config.go [hunk 2]
```diff
@@ -127,25 +129,37 @@ type ComponentConfig struct {
...
+		return err
 	}
-	ssc.StatefulSetName = ssn
-	if err := envconfig.Process("", &ssc); err != nil {
+	return fmt.Errorf("%q is not a valid stateful set controller ordinal", v)
+}
+
+// statefulSetConfig represents the required information for a StatefulSet service.
+type statefulSetConfig struct {
+	StatefulSetID statefulSetID `envconfig:"CONTROLLER_ORDINAL" required:"true"`
```
Comment (mattmoor) [inline leaderelection/config.go:150]: I wonder if we should make this `STATEFUL_ORDINAL` for consistency? 🤔

File: leaderelection/ordinal.go [hunk 3]
```diff
@@ -27,20 +27,17 @@ import (
 // should be set to the pod name via the downward API.
 const controllerOrdinalEnv = "CONTROLLER_ORDINAL"
 
-// ParseControllerOrdinal returns the parts of the StatefulSet's pod name.
-func ParseControllerOrdinal() (string, uint64, error) {
-	v := os.Getenv(controllerOrdinalEnv)
+// ControllerOrdinal tries to get ordinal from the pod name of a StatefulSet,
+// which is provided from the environment variable CONTROLLER_ORDINAL.
+func ControllerOrdinal() (int, error) {
```
Comment (mattmoor) [inline leaderelection/ordinal.go:32]: you really hate `uint` 🤣 

Why does this have to be public still?

File: leaderelection/config.go (hunk 2 above)
Comment (vagababov) [reply leaderelection/config.go:150]: Yeah, it was a sore in my eye a well. let's do it. Went for `STATEFUL_CONTROLLER_ORDINAL`.

File: leaderelection/ordinal.go [hunk 4]
```diff
@@ -27,20 +27,17 @@ import (
...
+	v, ok := os.LookupEnv(controllerOrdinalEnv)
+	if !ok {
+		return 0, fmt.Errorf("%s envvar is not set", controllerOrdinalEnv)
+	}
```
Comment (mattmoor) [inline leaderelection/ordinal.go:36]: Can we dedup the implementation with `newStatefulSetConfig`?

File: leaderelection/context.go [hunk 5]
```diff
@@ -53,10 +56,10 @@ func WithStandardLeaderElectorBuilder(ctx context.Context, kc kubernetes.Interfa
 	})
 }
 
-// WithStatefulSetLeaderElectorBuilder infuses a context with the ability to build
+// withStatefulSetElectorBuilder infuses a context with the ability to build
 // Electors which are assigned leadership based on the StatefulSet ordinal from
 // the provided component configuration.
-func WithStatefulSetLeaderElectorBuilder(ctx context.Context, cc ComponentConfig, ssc StatefulSetConfig) context.Context {
+func withStatefulSetElectorBuilder(ctx context.Context, cc ComponentConfig, ssc statefulSetConfig) context.Context {
```
Comment (yanweiguo) [inline leaderelection/context.go:62]: I think we might still want to make this public as autoscaler binary is not using the sharedmain. And only StatefulSet mode works for autoscaler. Falling back to Standard mode via `WithDynamicLeaderElectorBuilder` doesn't work in autoscaler case.

File: leaderelection/context.go [hunk 6]
```diff
@@ -36,10 +36,13 @@ import (
 // WithDynamicLeaderElectorBuilder sets up the statefulset elector based on environment,
 // falling back on the standard elector.
 func WithDynamicLeaderElectorBuilder(ctx context.Context, kc kubernetes.Interface, cc ComponentConfig) context.Context {
-	ssc, err := NewStatefulSetConfig()
+	logger := logging.FromContext(ctx)
+	ssc, err := newStatefulSetConfig()
 	if err == nil {
-		return WithStatefulSetLeaderElectorBuilder(ctx, cc, *ssc)
+		logger.Info("Running with StatefulSet leader election")
```
Comment (yanweiguo) [inline leaderelection/context.go:42]: I think the logs here are redundant as in `buildElector` we have similar logs.

File: leaderelection/ordinal.go (hunk 4 above)
Comment (vagababov) [reply leaderelection/ordinal.go:25]: done.

File: leaderelection/context.go (hunk 5 above)
Comment (vagababov) [reply leaderelection/context.go:62]: I talked with Matt about it a bit. We can just call WithDynaminLeaderElectorBuilder and make it fail if SS builder is expected.

File: leaderelection/context.go (hunk 6 above)
Comment (vagababov) [reply leaderelection/context.go:42]: 🤷  it's once per start log, so should be ok.

File: leaderelection/ordinal.go (hunk 3 above)
Comment (vagababov) [reply leaderelection/ordinal.go:21]: autoscaler will call it per @yanweiguo  :)

**Pull request merged**